mobile/android/android-device.ts

```diff
@@ -185,9 +188,166 @@ export class AndroidDevice implements Mobile.IDevice {
 			this.startPackageOnDevice(packageName).wait();
 			this.$logger.info("Successfully deployed on device with identifier '%s'", this.getIdentifier());
 		}).future<void>()();
-	}
-
-	private ensureFullAccessPermissions(devicePath: string): IFuture<void> {
+    }
+
+    private tcpForward(src: Number, dest: Number): IFuture<void> {
+        return (() => {
+            var tcpForwardCommand = this.composeCommand("forward tcp:%d tcp:%d", src.toString(), dest.toString());
+            this.$childProcess.exec(tcpForwardCommand).wait();
+        }).future<void>()();
+    }
+
+    private startDebuggerClient(port: Number): IFuture<void> {
+        return (() => {
+            var nodeInspectorModuleFilePath = require.resolve("node-inspector");
+            var nodeInspectorModuleDir = path.dirname(nodeInspectorModuleFilePath);
+            var nodeInspectorFullPath = path.join(path.join(nodeInspectorModuleDir, "bin"), "inspector");
+            this.$childProcess.spawn("node", [nodeInspectorFullPath, "--debug-port", port.toString()], { stdio: ["ignore", "ignore", "ignore"], detached: true });
+        }).future<void>()();
+    }
+
+    private openDebuggerClient(url: string): IFuture<void> {
+        return (() => {
+            this.$opener.open(url, "chrome");
+        }).future<void>()();
+    }
+
+    private printDebugPort(packageName: string): void {
+        var res = this.$childProcess.spawnFromEvent(this.adb, ["shell", "am", "broadcast", "-a", packageName + "-GetDgbPort"], "exit").wait();
+        this.$logger.info(res.stdout);
+    }
+
+    private attachDebugger(packageName: string): void {
+        var startDebuggerCommand = this.composeCommand("shell am broadcast -a \"%s-Debug\" --ez enable true", packageName);
+        var port = options["debug-port"];
+        if (port > 0) {
+            startDebuggerCommand += " --ei debuggerPort " + options["debug-port"];
+            this.$childProcess.exec(startDebuggerCommand).wait();
+        } else {
+            var res = this.$childProcess.spawnFromEvent(this.adb, ["shell", "am", "broadcast", "-a", packageName + "-Debug", "--ez", "enable", "true"], "exit").wait();
+            var match = res.stdout.match(/result=(\d)+/);
+            if (match) {
+                port = match[0].substring(7);
+            } else {
+                port = 0;
+            }
+        }
+        if (port > 0) {
+            this.tcpForward(port, port).wait();
+        }
+        this.startDebuggerClient(port).wait();
```

Comment:
At this moment, it is possible for port to be 0. Should we call startDebugger()?